MiniSpace.Services.Posts/src/MiniSpace.Services.Posts.Application/Commands/Handlers/ChangePostStateHandler.cs

```diff
@@ -0,0 +1,86 @@
+
+namespace MiniSpace.Services.Posts.Application.Commands.Handlers
+{
+    public class ChangePostStateHandler : ICommandHandler<ChangePostState>
+    {
+        private readonly IPostRepository _postRepository;
+        private readonly IAppContext _appContext;
+        private readonly IDateTimeProvider _dateTimeProvider;
+        private readonly IMessageBroker _messageBroker;
+        
+        public ChangePostStateHandler(IPostRepository postRepository, IAppContext appContext,
+            IDateTimeProvider dateTimeProvider, IMessageBroker messageBroker)
+        {
+            _postRepository = postRepository;
+            _appContext = appContext;
+            _dateTimeProvider = dateTimeProvider;
+            _messageBroker = messageBroker;
+        }
+        
+        public async Task HandleAsync(ChangePostState command, CancellationToken cancellationToken = default)
+        {
+            var post = await _postRepository.GetAsync(command.PostId);
+            if (post is null)
+            {
+                throw new PostNotFoundException(command.PostId);
+            }
+            
+            var identity = _appContext.Identity;
+            if (identity.IsAuthenticated && identity.Id != post.StudentId && !identity.IsAdmin)
+            {
+                throw new UnauthorizedPostAccessException(command.PostId, identity.Id);
+            }
+            
+            if (!Enum.TryParse<State>(command.State, true, out var newState))
+            {
+                throw new InvalidPostStateException(command.State);
+            }
+
+            if (!identity.IsAdmin && post.State == State.Reported)
+            {
+                throw new UnauthorizedPostOperationException(command.PostId, identity.Id);
+            }
+            
+            if (post.State == newState && post.State != State.ToBePublished)
+            {
+                throw new PostStateAlreadySetException(post.Id, newState);
+            }
+            
+            var previousState = post.State.ToString().ToLowerInvariant();
+
+            switch (newState)
+            {
+                case State.ToBePublished:
+                    post.SetToBePublished(command.PublishDate
+                                          ?? throw new PublishDateNullException(command.PostId, newState),
+                        _dateTimeProvider.Now);
+                    break;
+                case State.Published:
+                    post.SetPublished();
+                    break;
+                case State.InDraft:
+                    post.SetInDraft();
+                    break;
+                case State.Hidden:
+                    post.SetHidden();
+                    break;
+                case State.Reported:
+                    post.SetReported();
+                    break;
+                default:
+                    throw new InvalidPostStateException(post.State.ToString().ToLowerInvariant());
+            }
+            
```

Comment:
Hmmm, maybe strategy?

Reply:
We used to switch there through states in other services also so I think there is no necessity to reorganize it.